src/System.IO.Pipelines/Pipe.cs

```diff
@@ -343,15 +357,21 @@ public void AdvanceWriter(int bytesWritten)
             } // and if zero, just do nothing; don't need to validate tail etc
         }
 
-        internal async Task<bool> FlushAsync()
+        internal WritableBufferAwaitable FlushAsync()
         {
             if (_producingState == State.Active)
             {
                 // Commit the data as not already committed
                 Commit();
             }
 
-            return await SignalWriterAsync();
+            // TODO: Can factor out this lock
+            lock (_sync)
```

Comment:
Just remove this lock, I don't think there is anything in Resume to protect and running callback via inline scheduler would cause unnecessary overlocking.